Update saving / loading checkpoint documentation

jwyyy opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR updates saving / loading checkpoint examples. 

## Related issue number

Closes #20910.

<!-- Please give a short summary of the change and the problem this solves. -->

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -540,26 +540,61 @@ equivalent to the sum of the number of epochs.
+            train.save_checkpoint(epoch=epoch, model_weights=model.state_dict())
+
+
+    trainer = Trainer(backend="torch", num_workers=1)
```
Can you set `num_workers=2` so that this is more representative of distributed training?

matthewdeng replied:
This should also include wrapping the model in `train.torch.prepare_model`, and calling `consume_prefix_in_state_dict_if_present` to support transferable saving & loading (see https://github.com/ray-project/ray/issues/20915).

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -540,26 +540,61 @@ equivalent to the sum of the number of epochs.
 
     trainer.run(train_func, config={"num_epochs": 1})
     print(trainer.latest_checkpoint)
-    # {'epoch': 0, 'model': 0}
+    # {'epoch': 0,
+    # 'model_weights': OrderedDict([('weight', tensor([[-0.4763, -0.3775,  0.4768,  0.3969]])),
+    #                               ('bias', tensor([0.2237]))]),
+    # '_timestamp': 1638979904}
 
     trainer.run(train_func, config={"num_epochs": 5})
     print(trainer.latest_checkpoint)
-    # {'epoch': 4, 'model': 10}
+    # {'epoch': 4,
+    # 'model_weights': OrderedDict([('weight', tensor([[ 0.4854, -0.1520, -0.1217,  0.2099]])),
+    #                               ('bias', tensor([-0.0724]))]),
+    # '_timestamp': 1638979904}
+
```
We can remove these lines altogether, looking at it again now I don't think they add much value to the example.

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -540,26 +540,61 @@ equivalent to the sum of the number of epochs.
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            # For TensorFlow (Keras) models, use model.get_weights()
```
Thanks for leaving this comment here. In the long run we should have individual tabs showing a concrete example for each backend. Could you start by just adding a tab for PyTorch?

```
.. tabs::
  .. group-tab:: PyTorch
```

jwyyy replied:
@matthewdeng Thank you for your feedback! I think I can add a TF example. Revising other parts as well.

matthewdeng replied:
Thanks 💯 for adding the TF example as well!

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -540,26 +540,61 @@ equivalent to the sum of the number of epochs.
     from ray import train
     from ray.train import Trainer
 
+    import torch
```
Adding a comment here, can you update the text above which is no longer True?

```
As a trivial example, let's create a "model" which is an integer value that is
equivalent to the sum of the number of epochs.
```

matthewdeng reviewed (changes requested):
Thanks for picking this up! Added a few initial comments, let me know if you have any questions.

jwyyy commented:
Hi @matthewdeng, thank you for your feedback! I have updated the PR and added TF examples. Please let me know your comments after you have time to review it again. Thanks!

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,100 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
+            # create a toy dataset
+            # data   : X - dim = (n, 4)
+            # target : Y - dim = (n, 1)
+            Y = torch.Tensor(np.random.uniform(0, 1, size=(n, 1)))
+            X = torch.Tensor(np.random.normal(0, 1, size=(n, 4)))
```
nit: Reorder these so X comes before Y for readability.

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,100 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
+            # toy neural network : 1-layer
+            # wrap the model in DDP
+            model = ray.train.torch.prepare_model(nn.Linear(4, 1))
+            mse = nn.MSELoss()
```
nit: Rename this to `criterion` so it's more general.

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,100 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
+                # w/o DDP: model.state_dict()
+                # w/  DDP: model.module.state_dict()
+                # See: https://github.com/ray-project/ray/issues/20915
+                ddp_state_dict = model.state_dict()
```
nit: Rename this to `state_dict` since this should also work if the model is not wrapped in DDP (when `num_workers=1`).

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,100 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
-    trainer.run(train_func, config={"num_epochs": 3})
-    print(trainer.latest_checkpoint)
-    # {'epoch': 2, 'model': 3}
+        print(trainer.latest_checkpoint)
```
Can you include a sample output here?

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,100 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
+            with strategy.scope():
+                # toy neural network : 1-layer
+                model = tf.keras.Sequential([tf.keras.layers.Dense(1, activation="linear")])
+                model.compile(loss="mean_squared_error", metrics=["mse"])
```
Nit for consistency with the PyTorch example: 
```suggestion
                model.compile(optimizer="Adam", loss="mean_squared_error", metrics=["mse"])
```

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,100 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
+                model.compile(loss="mean_squared_error", metrics=["mse"])
+
+            for epoch in range(config["num_epochs"]):
+                model.fit(X, Y, epochs=epoch, batch_size=20)
```
We should only call fit once since we're iterating through the epochs already!
```suggestion
                model.fit(X, Y, batch_size=20)
```

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -652,28 +716,127 @@ Checkpoints can be loaded into the training function in 2 steps:
 
-    from ray import train
-    from ray.train import Trainer
+  .. group-tab:: PyTorch
```
Suggestion: add `:emphasize-lines: <line numbers>` here to highlight the lines added.

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -652,28 +716,127 @@ Checkpoints can be loaded into the training function in 2 steps:
+            checkpoint = train.load_checkpoint()
+            if checkpoint:
+                # assume that we have run the train.save_checkpoint() example
+                #                and successfully save some model weights
```
nit: Spacing looks weird here.
```suggestion
                # and successfully save some model weights
```

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -652,28 +716,127 @@ Checkpoints can be loaded into the training function in 2 steps:
+
+            with strategy.scope():
+                # toy neural network : 1-layer
+                model = tf.keras.Sequential([tf.keras.layers.Dense(1, activation="linear")])
+                checkpoint = train.load_checkpoint()
+                if checkpoint:
+                    model.build((None, 4))
```
You can get rid of the `build` step by adding `input_shape` - I think you need this for the non-checkpoint flow as well!

```suggestion
                model = tf.keras.Sequential([tf.keras.layers.Dense(1, activation="linear", input_shape=(4,))])
                checkpoint = train.load_checkpoint()
                if checkpoint:
```

jwyyy replied:
Hi @matthewdeng, thank you for your suggestion! It looks better to add input shape. I will update the example shortly.

For non-checkpoint flow, the input size will be inferred when we call `fit()`, `eval()` and `predict()` for the first time (see the last example in [Examples](https://www.tensorflow.org/api_docs/python/tf/keras/Sequential#examples_3)).

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -652,28 +716,127 @@ Checkpoints can be loaded into the training function in 2 steps:
+        trainer.run(train_func, config={"num_epochs": 2})
+        trainer.shutdown()
+
+        # restart the trainer for the loading checkpoint example
+        # TensorFlow ops need to be created after a MultiWorkerMirroredStrategy instance is created.
+        # See: https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras#train_the_model_with_multiworkermirroredstrategy
```
Thanks for documenting this!

matthewdeng reviewed (commented):
This looks great, thanks for a ton for adding this! Mostly just minor comments this time.

jwyyy commented:
Hi @matthewdeng, thank you for your review! I have made some changes according to your feedback. All examples were tested locally. Please let me know what your comments are. Thanks!

jwyyy commented:
Could you also take a look at the failed tests? The bootstrap env build test was okay in previous commits. Since the changes in this PR are in doc not core code, I am not sure why the build test failed. Please let me know how I can fix them. Thanks!

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,105 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
-            model += epoch # This should be replaced with training logic.
-            train.save_checkpoint(epoch=epoch, model=model)
+    .. code-block:: python
+        :emphasize-lines: 21, 37, 38, 39
```
No need to highlight `prepare_model` here.
```suggestion
        :emphasize-lines: 37, 38, 39
```

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -532,36 +532,105 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
+  .. group-tab:: TensorFlow
+
+    .. code-block:: python
+        :emphasize-lines: 16, 17, 19, 20, 24
```
No need to highlight strategy/model compilation.
```suggestion
        :emphasize-lines: 24
```

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -652,28 +721,131 @@ Checkpoints can be loaded into the training function in 2 steps:
-            model += epoch
-            train.save_checkpoint(epoch=epoch, model=model)
+    .. code-block:: python
+        :emphasize-lines: 26, 30, 34
```
Similar as above, emphasize the changed lines (`start_epoch` and loading).
```suggestion
        :emphasize-lines: 24, 26, 27, 30, 31, 35
```

matthewdeng commented on doc/source/train/user_guide.rst:
```diff
@@ -652,28 +721,131 @@ Checkpoints can be loaded into the training function in 2 steps:
+  .. group-tab:: TensorFlow
+
+    .. code-block:: python
+        :emphasize-lines: 22, 26
```
```suggestion
        :emphasize-lines: 16, 22, 23, 26, 27, 30
```

matthewdeng reviewed (approved):
Overall looks good to me! GPU environment build task is unrelated and isn't a dependency for any doc-related tests. 

@amogkam can you take a pass over this?

jwyyy commented:
@matthewdeng, thank you for your review and updates!

amogkam reviewed (approved):
Thanks for doing this! Looks great!
